include/common/mir/input/keymap.h

```diff
@@ -20,58 +20,45 @@
 #ifndef MIR_INPUT_KEYMAP_H_
 #define MIR_INPUT_KEYMAP_H_
 
+#include <memory>
 #include <string>
-#include <ostream>
+
+struct xkb_keymap;
+struct xkb_context;
 
 namespace mir
 {
 namespace input
 {
+using XKBKeymapPtr = std::unique_ptr<xkb_keymap, void(*)(xkb_keymap*)>;
 
-struct Keymap
+class Keymap
 {
+public:
     Keymap() = default;
-    Keymap(std::string&& model,
-           std::string&& layout,
-           std::string&& variant,
-           std::string&& options)
-        : model{model}, layout{layout}, variant{variant}, options{options}
-    {
-    }
-
-    Keymap(std::string const& model,
-           std::string const& layout,
-           std::string const& variant,
-           std::string const& options)
-        : model{model}, layout{layout}, variant{variant}, options{options}
-    {
-    }
+    virtual ~Keymap() = default;
 
-    std::string model{"pc105+inet"};
-    std::string layout{"us"};
-    std::string variant;
-    std::string options;
+    /// If other's type is the same as or derived from this's type, and it's values are equal
+    virtual auto matches(Keymap const& other) const -> bool = 0;
```

Comment:
OK, I know this is how I described the initial implementation. But, still on the "what does it actually mean" theme I don't think this is what is actually needed by the code that uses it (via `operator==()`).

There are three uses:
1. `mi::DefaultDevice::set_keyboard_configuration()` actually wants to know if the `XkbMappingState` has changed. It really doesn't care about the type.
2. `mf::WlSeat::ConfigObserver::device_added()/device_changed()/changes_complete()` wants to know about the equivalence of a potential call to `make_unique_xkb_keymap()`.
3. `MirKeyboardConfig::operator==()` is used by `MirInputDevice::operator==()` where it might be valid. I don't think the "different objects with the same dynamic type and value" is a significant case. Just "same object" looks like enough.

I think trying to make all these uses fit an equivalence relationship is confusing.